src/fides/api/ops/service/privacy_request/consent_email_batch_service.py

```diff
@@ -0,0 +1,260 @@
+    privacy_requests: Query, batched_user_data: List[BatchedUserConsentData]
+) -> None:
+    """
+    Collect user identities and consent preferences across privacy requests for each applicable connector
+
+    ! Edits batched_user_data in place
+    """
+    for privacy_request in privacy_requests:
+        user_identities: Dict[str, Any] = privacy_request.get_cached_identity_data()
+
+        for pending_email in batched_user_data:
+            filtered_user_identities: Dict[
+                str, Any
+            ] = filter_user_identities_for_connector(
+                pending_email.connection_secrets, user_identities
+            )
+
+            if filtered_user_identities and privacy_request.consent_preferences:
+                pending_email.batched_user_consent_preferences.append(
+                    ConsentPreferencesByUser(
+                        identities=filtered_user_identities,
+                        consent_preferences=privacy_request.consent_preferences,
+                    )
+                )
+            else:
+                pending_email.skipped_privacy_requests.append(privacy_request.id)
+
+
+def send_prepared_emails(
+    db: Session,
+    batched_user_data: List[BatchedUserConsentData],
+    privacy_requests: Query,
+) -> int:
+    """Send a single consent email for each connector using the prepared data in batched_user_data
+
+    Also add audit logs to each relevant privacy request.
+    """
+    emails_sent: int = 0
+    for pending_email in batched_user_data:
+        if not pending_email.batched_user_consent_preferences:
+            logger.info(
+                "Skipping consent email send for connector: '{}'. "
+                "No corresponding user identities found for pending privacy requests.",
+                pending_email.connection_name,
+            )
+            continue
+
+        logger.info(
+            "Sending batched consent email for connector {}...",
+            pending_email.connection_name,
+        )
+        send_single_consent_email(
+            db=db,
+            subject_email=pending_email.connection_secrets.recipient_email_address,
+            subject_name=pending_email.connection_secrets.third_party_vendor_name,
+            required_identities=pending_email.required_identities,
+            user_consent_preferences=pending_email.batched_user_consent_preferences,
+            test_mode=False,
+        )
+
+        for privacy_request in privacy_requests:
+            if privacy_request.id not in pending_email.skipped_privacy_requests:
+                AuditLog.create(
+                    db=db,
+                    data={
+                        "user_id": "system",
+                        "privacy_request_id": privacy_request.id,
+                        "action": AuditLogAction.email_sent,
+                        "message": f"Consent email instructions dispatched for '{pending_email.connection_name}'",
+                    },
+                )
```

Comment:
In this email batch service, I do a lot of looping through the whole list of privacy requests multiple times.  I do things like collecting all the relevant user data from all the privacy requests, and then here, adding auditlogs where applicable, and later, updating the privacy request status on each and requeuing.

My opinion is that this is fine for now with the number of privacy requests and consent-email type connectors supported, but at some point we might need to improve this - for example, we might collect all the audit logs we need to make for all privacy requests for all email connector types that succeeded as we go and do one bulk insert of the audit logs into the db.